cmd/restic/cmd_rewrite.go

```diff
@@ -0,0 +1,296 @@
+		}
+		opts.Excludes = append(opts.Excludes, excludes...)
+	}
+
+	if len(opts.InsensitiveExcludes) > 0 {
+		fs = append(fs, rejectByInsensitivePattern(opts.InsensitiveExcludes))
+	}
+
+	if len(opts.Excludes) > 0 {
+		fs = append(fs, rejectByPattern(opts.Excludes))
+	}
+
+	return fs, nil
+}
+
+func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *restic.Snapshot, opts RewriteOptions, gopts GlobalOptions) (bool, error) {
+	if sn.Tree == nil {
+		return false, errors.Errorf("snapshot %v has nil tree", sn.ID().Str())
+	}
+
+	var saveTreeFunc saveTreeFunction
+	if !opts.DryRun {
+		saveTreeFunc = func(tree *restic.Tree) (restic.ID, error) {
+			return repo.SaveTree(ctx, tree)
+		}
+	} else {
+		saveTreeFunc = func(tree *restic.Tree) (restic.ID, error) {
+			return restic.ID{0}, nil
+		}
+	}
+
+	rejectByNameFuncs, err := collectRejectFuncsForRewrite(opts)
+	if err != nil {
+		return false, err
+	}
+
+	checkExclude := func(nodepath string) bool {
+		for _, reject := range rejectByNameFuncs {
+			if reject(nodepath) {
+				return true
+			}
+		}
+		return false
+	}
+
+	filteredTree, err := filterNode(ctx, repo, "/", *sn.Tree, checkExclude, saveTreeFunc)
+
+	if err != nil {
+		return false, err
+	}
+
+	if filteredTree == *sn.Tree {
+		debug.Log("Snapshot not touched\n")
+		return false, nil
+	}
+
+	debug.Log("Snapshot modified\n")
+	if opts.DryRun {
+		Printf("Will modify snapshot: %s\n", sn.String())
+		return true, nil
+	}
+
+	err = repo.Flush(ctx)
+	if err != nil {
+		return false, err
+	}
+	err = repo.SaveIndex(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	// Retain the original snapshot id over all tag changes.
+	if sn.Original == nil {
+		sn.Original = sn.ID()
+	}
+	*sn.Tree = filteredTree
+
+	if !opts.Inplace {
+		sn.AddTags([]string{"rewrite"})
+	}
+
+	// Save the new snapshot.
+	id, err := repo.SaveJSONUnpacked(ctx, restic.SnapshotFile, sn)
+	if err != nil {
+		return false, err
+	}
+	err = repo.Flush(ctx)
+	if err != nil {
+		return true, err
+	}
+
+	if opts.Inplace {
+		h := restic.Handle{Type: restic.SnapshotFile, Name: sn.ID().String()}
+		if err = repo.Backend().Remove(ctx, h); err != nil {
+			return false, err
+		}
+
+		debug.Log("old snapshot %v removed", sn.ID())
+	}
+	Printf("new snapshot saved as %v\n", id)
+	return true, nil
+}
+
+func runRewrite(opts RewriteOptions, gopts GlobalOptions, args []string) error {
+
+	if len(opts.Hosts) == 0 && len(opts.Tags) == 0 && len(opts.Paths) == 0 && len(args) == 0 {
+		return errors.Fatal("no snapshots provided")
+	}
+
+	if len(opts.ExcludeFiles) == 0 && len(opts.Excludes) == 0 && len(opts.InsensitiveExcludes) == 0 {
+		return errors.Fatal("Nothing to do: no excludes provided")
+	}
+
+	if len(args) == 1 && args[0] == "all" {
+		args = []string{}
+	}
+
+	repo, err := OpenRepository(gopts)
+	if err != nil {
+		return err
+	}
+
+	if !gopts.NoLock && !opts.DryRun {
+		Verbosef("create exclusive lock for repository\n")
+		lock, err := lockRepoExclusive(repo)
```

Comment:
If you outsource the deletion of files, a non-exclusive lock would suffice here.

Reply:
Upgrading a lock in the middle of a command is a problem. There is no guarantee that this will work.